README.md
# RemoteCompiler

remotecompiler/CHANGELOG.md
# Change Log
All notable changes to the "remotecompiler" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]
- Initial release

remotecompiler/src/config.ts
/*
 * @Author: PLUS 
 * @Date: 2018-10-30 20:19:37 
 * @Last Modified by: PLUS
 * @Last Modified time: 2018-10-30 20:20:09
 */

import {
    WorkspaceConfiguration,
    window
} from 'vscode';

import * as child from 'child_process';

export class Config {
    rsync_exe: string;
    ssh_exe:string;
    shell: string;
    project_name: string;
    remote_port: string;
    remote_host: string;
    remote_path: string;
    local_path: string;
    onSaveIndividual: boolean;
    cygpath: string;
    executableshell: string;
    reg_exp:string;

    excutecmd: string;
    makefilepath: string;
    is_use_remote_backup: boolean;
    remote_host_bk: string;
    remote_path_bk: string;
    remote_port_bk: string;

    constructor(config: WorkspaceConfiguration) {
        this.rsync_exe = config.get('rsync');
        this.ssh_exe = config.get('ssh');
        this.shell = config.get('shell');
        this.cygpath = config.get('cygpath', "");
        this.project_name = config.get('project_name', "");
        this.remote_port = config.get('remote_port', "");
        this.remote_host = config.get('remote_host', "");
        this.remote_path = config.get('remote_path', "");
        this.local_path = config.get('local_path', "");
        this.onSaveIndividual = config.get('onSaveIndividual', false);
        this.executableshell = config.get('executableShell');
        this.excutecmd = config.get("excutecmd");
        this.makefilepath = config.get("makefilepath");
        this.reg_exp = config.get('reg_exp',"");

        this.remote_port_bk = config.get('remote_port_bk', "");
        this.remote_path_bk = config.get('remote_path_bk', "");
        this.remote_host_bk = config.get('remote_host_bk', "");
        this.is_use_remote_backup = config.get("is_use_remote_backup", false);
    }

    translatePath(path: string): string {
        if (this.cygpath) {
            let rtn = child.spawnSync(this.cygpath, [path]);
            if (rtn.status !== 0) {
                throw new Error("Path Tranlate Issue");
            }

            if (rtn.error) {
                throw rtn.error;
            }

            let s_rtn = rtn.stdout.toString();
            s_rtn = s_rtn.trim();
            return s_rtn;
        }

        return path;

    }

}

remotecompiler/src/CmdBase.ts
/*
 * @Author: PLUS 
 * @Date: 2018-10-30 13:54:21 
 * @Last Modified by: PLUS
 * @Last Modified time: 2018-10-30 20:22:56
 */

'use strict';
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below

//require "../node_modules/@types/rsync/index.d.ts";
/// <reference path="../node_modules/@types/rsync/index.d.ts">
import * as vscode from 'vscode';
import * as child from 'child_process';

let currentSync: child.ChildProcess;

export abstract class CmdBase {
    protected out: vscode.OutputChannel;

    constructor(out: vscode.OutputChannel) {
        this.out = out;
    }

    handle_error(err: { code: string, message: string }): void {
        this.out.appendLine("ERROR > " + err.message);
    }

    handle_stdout(data: Buffer): void {
        this.out.append(data.toString());
    }

    handle_stderr(data: Buffer): void {
        this.out.append(data.toString());
    }

    execute(cmd: string, args: string[] = [], shellpro: string = ""): Promise<number> {
        return new Promise<number>(resolve => {
            let is_error: boolean = false;

            this.out.appendLine(`> ${cmd} ${args.join(" ")} `);

            if (process.platform === 'win32' && shellpro) {
                // when the platform is win32, spawn would add /s /c flags, making it impossible for the 
                // shell to be something other than cmd or powershell (e.g. bash)
                args = ["'", cmd].concat(args, "'");
                this.out.appendLine("platform:" + process.platform + ", shell:" + shellpro + ", args:" + args.join(" "));
                currentSync = child.spawn(shellpro + " -s -c", args, { stdio: 'pipe', shell: "cmd.exe" });
            } else {
                currentSync = child.spawn(cmd, args, { stdio: 'pipe', shell: shellpro });
            }

            currentSync.on('error', function (err: { code: string, message: string }) {
                this.handle_error(err);
                is_error = true;
                resolve(1);
            });

            let showStdOut = (data: Buffer): void => {
                this.handle_stdout(data);
            };

            let showStdOutEnd = (): void => {
                this.out.appendLine(`${cmd} stdout completed`);
            };
            currentSync.stdout.on('data', showStdOut);
            currentSync.stdout.on('end', showStdOutEnd);
            
            let showStdErr = (data: Buffer): void => {
                this.handle_stderr(data);
            };
            let showStdErrEnd = (): void => {
                this.out.appendLine(`${cmd} stderr completed`);
            };
            currentSync.stderr.on('data', showStdErr);
            currentSync.stderr.on('end', showStdErrEnd);
            // currentSync.stderr.on('data', function (data: Buffer): void {
            //     this.handle_stderr(data);
            // });

            currentSync.on('close', function (code) {
                //this.out.appendLine("currentSync close");
                if (is_error) {
                    return;
                }
                resolve(code);
            });
        });
    }

}



remotecompiler/src/make_cmd.ts
/*
 * @Author: PLUS 
 * @Date: 2018-10-30 20:20:41 
 * @Last Modified by: PLUS
 * @Last Modified time: 2018-11-05 18:56:55
 */
import * as vscode from 'vscode';
import * as path from 'path';
import { CmdBase } from './CmdBase';

export class MakeCmd extends CmdBase {
    private curDir: string;
    private remotePath: string;
    private localPath: string;
    private dc: vscode.DiagnosticCollection;
    private regExp: string;
    private makefilepath: string;

    constructor(out: vscode.OutputChannel,
        localPath: string,
        remotePath: string,
        makefilepath: string,
        regExp: string,
        dc: vscode.DiagnosticCollection) {
        super(out);
        this.localPath = localPath;
        this.remotePath = remotePath;
        this.makefilepath = makefilepath;
        this.regExp = regExp;
        this.dc = dc;
    }

    ParseStdOut(str: string): void {
        // make[1]: Entering directory `/data/williamyjou/rsyncdemo/app_common'
        const rOutputGcc: RegExp = new RegExp("^(.*): (Entering directory) `(.*)'$");
        let j: number = 0;
        let strs: string[] = str.split("\n");
        strs.forEach(textLine => {
            if (textLine) {
                // this.out.appendLine("Info[" + j + "]:" + textLine);
                // parse warning and error from the given text line
                let matches: RegExpExecArray = rOutputGcc.exec(textLine);

                if (matches && matches.length >= 4) {
                    // get warning and error info
                    let tag: string = "";

                    let action: string = "Entering directory";

                    let filepath: string = "";

                    tag = matches[1].trim();

                    action = matches[2].trim();

                    filepath = matches[3].trim();

                    let fileout = filepath.split(this.remotePath);
                    this.out.appendLine("fileout:" + fileout[1]);

                    this.curDir = fileout[1];
                }
            }
            j++;
        });
    }

    ParseStdErr(str: string): void {
        // init regex of gcc/clang output
        //middle_service.cpp:30: error: 'struct TRPC_SVCINFO' has no member named 'olen1'
        //this.out.appendLine("str:"+str);

        // const rOutputGcc: RegExp = new RegExp("^(.*):([0-9]*):([0-9]*): (.*错误：|.*error:)(.*)$");
        // const rOutputGcc: RegExp = new RegExp("^(.*):([0-9]*): (.*error:)(.*)$");
        this.out.appendLine(this.regExp);
        const rOutputGcc: RegExp = new RegExp(this.regExp);
        // init diagnostics map
        let diagnosticsMap: Map<string, vscode.Diagnostic[]> = new Map();

        let j: number = 0;
        let strs: string[] = str.split("\n");
        strs.forEach(textLine => {
            if (textLine) {
                this.out.appendLine("Error[" + j + "]:" + textLine);
                // parse warning and error from the given text line
                let matches: RegExpExecArray = rOutputGcc.exec(textLine);

                if (matches && matches.length >= 5) {
                    // get warning and error info
                    let file: string = "";

                    let line: string = "0";

                    let column: string = "0";

                    let kind: string = "error";

                    let message: string = "";

                    file = matches[1].trim();

                    line = matches[2].trim();

                    //  column = matches[3].trim();

                    kind = matches[3].toLocaleLowerCase().trim();

                    message = matches[4].trim();
                    let uri: vscode.Uri;
                    // get uri of file
                    if (this.curDir !== undefined) {
                        uri = vscode.Uri.file(path.isAbsolute(file) ? file : path.join(this.localPath + "\\" + this.makefilepath + "\\" + this.curDir, file));
                    }
                    else {
                        uri = vscode.Uri.file(path.isAbsolute(file) ? file : path.join(this.localPath + "\\" + this.makefilepath, file));
                    }

                    this.out.appendLine("uri.fsPath:" + uri.fsPath);

                    // init severity 
                    let severity: vscode.DiagnosticSeverity = { error: vscode.DiagnosticSeverity.Error, warning: vscode.DiagnosticSeverity.Warning }[kind];

                    if (severity !== vscode.DiagnosticSeverity.Error && severity !== vscode.DiagnosticSeverity.Warning) {
                        severity = vscode.DiagnosticSeverity.Error;
                    }
                    // get start line and column
                    let startLine: number = Number(line);
                    let startColumn: number = Number(column);

                    startLine = startLine > 0 ? startLine - 1 : startLine;
                    startColumn = startColumn > 0 ? startColumn - 1 : startColumn;

                    // get end line and column
                    let endLine: number = startLine;
                    let endColumn: number = startColumn;

                    // init range
                    let range = new vscode.Range(startLine, startColumn, endLine, endColumn);
                    // save diagnostic
                    let diagnostic: vscode.Diagnostic = new vscode.Diagnostic(range, message, severity);

                    // get diagnostics of this file
                    let diagnostics: vscode.Diagnostic[] = diagnosticsMap.get(uri.fsPath);

                    // init diagnostics
                    if (!diagnostics) {
                        diagnostics = [];
                        diagnosticsMap.set(uri.fsPath, diagnostics);
                    }

                    // add diagnostic to diagnostics
                    diagnostics.push(diagnostic);
                }
            }
            j++;
        });

        diagnosticsMap.forEach((diagnostics: vscode.Diagnostic[], fsPath: string) => {
            this.dc.set(vscode.Uri.file(fsPath), diagnostics);
        });

        diagnosticsMap.clear();
    }

    handle_error(err: { code: string, message: string }): void {
        this.out.appendLine("ERROR > " + err.message);
    }

    handle_stdout(data: Buffer): void {
        this.out.append(data.toString());
        let str: string = data.toString();
        this.ParseStdOut(str);
    }

    handle_stderr(data: Buffer): void {
        this.out.append(data.toString());
        let str: string = data.toString();
        this.ParseStdErr(str);
    }
}

remotecompiler/src/extension.ts
/*
 * @Author: PLUS 
 * @Date: 2018-10-30 20:20:33 
 * @Last Modified by: PLUS
 * @Last Modified time: 2018-10-30 20:50:37
 */
'use strict';
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below

//require "../node_modules/@types/rsync/index.d.ts";
/// <reference path="../node_modules/@types/rsync/index.d.ts">
import * as vscode from 'vscode';
import * as child from 'child_process';
import { Config } from './config';
import Rsync = require("rsync");
import { RsyncCmd } from './rsync_cmd';
import { MakeCmd } from './make_cmd';

/*
 * Set up the command using the fluent interface, starting with an
 * empty command wrapper and adding options using methods.
 */

// short name
let vswin = vscode.window;

// view compent decalare
export const outputChannel: vscode.OutputChannel = vswin.createOutputChannel('RemoteCompiler');
const createStatusText = (text: string): string => `RemoteCompiler: ${text}`;
const statusbar: vscode.StatusBarItem = vswin.createStatusBarItem(vscode.StatusBarAlignment.Right, 1);

// child process
let currentSync: child.ChildProcess = <child.ChildProcess>{};
const getConfig = (): Config => new Config(vscode.workspace.getConfiguration('remotecompiler'));

let curDir: string = "";

let dc: vscode.DiagnosticCollection = vscode.languages.createDiagnosticCollection();

const runSync = function (isDryRun: boolean, args: string[]): Promise<number> {
    outputChannel.clear();

    const syncStartTime: Date = new Date();
    outputChannel.appendLine(`\n${syncStartTime.toString()} ${isDryRun ? 'comparing' : 'syncing'}`);

    let rsync_cmd: RsyncCmd = new RsyncCmd(outputChannel);
    let rsync_exe: string = getConfig().rsync_exe;
    let executableshell: string = getConfig().executableshell;
    return rsync_cmd.execute(rsync_exe, args, executableshell);
};

const runMake = function (args: string[]): Promise<number> {
    dc.clear();

    const syncStartTime: Date = new Date();
    outputChannel.appendLine(`\n${syncStartTime.toString()}'making'`);

    let make_cmd: MakeCmd = new MakeCmd(outputChannel,
        getConfig().local_path,
        getConfig().remote_path,
        getConfig().makefilepath,
        getConfig().reg_exp,
        dc);
    let ssh_exe: string = getConfig().ssh_exe;
    let executableshell: string = getConfig().executableshell;
    return make_cmd.execute(ssh_exe, args, executableshell);
};

const RemoteCompile = async function (): Promise<void> {
    outputChannel.show();

    statusbar.color = 'mediumseagreen';
    outputChannel.appendLine("Enter sync");
    statusbar.text = createStatusText('Syncing');

    let config: Config = getConfig();

    let rsync_ins = new Rsync();
    let success = true;

    let paths = [];

    let localpath: string = config.translatePath(config.local_path + "\\" + config.project_name);
    // let remotepath: string = "williamyjou@10.123.2.156:/data/williamyjou/rsyncdemo/";

    let remote_host: string = config.remote_host;
    let remote_path: string = config.remote_path;
    let remote_port: string = config.remote_port;
    if (config.is_use_remote_backup) {
        remote_host = config.remote_host_bk;
        remote_path = config.remote_path_bk;
        remote_port = config.remote_port_bk;
    }

    let remotehostpath: string = config.translatePath(remote_host) + ":" + config.translatePath(remote_path);
    paths = [localpath, remotehostpath];

    rsync_ins = rsync_ins
        .flags("vczrtu")
        .progress();

    rsync_ins = rsync_ins.exclude([".vscode", ".svn", ".git"]);

    rsync_ins = rsync_ins.shell(config.shell + " " + remote_port);

    let rtn = await runSync(rsync_ins.isSet('n'), rsync_ins.args().concat(paths));
    if (rtn === 0) {
        success = success && true;
    } else {
        vswin.showErrorMessage("rsync return " + rtn);
        success = false;
    }

    if (!success) {
        statusbar.color = 'red';
        statusbar.text = createStatusText('Rsync error');
        statusbar.show();
        return;
    }

    statusbar.color = undefined;
    statusbar.text = createStatusText('Sync Succ');
    statusbar.show();

    outputChannel.appendLine("Sync Complete");
    vswin.showInformationMessage("Sync Complete");

    outputChannel.appendLine("Making");
    let excute_cmd = "-p " 
                + remote_port + " " 
                + remote_host + " \"cd " 
                + remote_path + "/" 
                + config.project_name + "/"
                + config.makefilepath + " && " 
                + config.excutecmd +"\"";
    let args: string[] = [excute_cmd];
    rtn = await runMake(args);
    if (rtn === 0) {
        success = success && true;
        outputChannel.appendLine("Make Compelte");
        statusbar.text = "Make Compelte";
        vswin.showInformationMessage("Make Complete");
    } else {
        vswin.showErrorMessage("AfterSync return " + rtn);
    }
};

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {
    // Use the console to output diagnostic information (console.log) and errors (console.error)
    // This line of code will only be executed once when your extension is activated
    console.log('Congratulations, your extension "Remote Compiler" is now active!');
    // The command has been defined in the package.json file
    // Now provide the implementation of the command with  registerCommand
    // The commandId parameter must match the command field in package.json

    let remotecompile: vscode.Disposable = vscode.commands.registerCommand('RemoteCompile.Execute', () => {
        outputChannel.appendLine("Remote Compiling...");
        RemoteCompile();
    });

    context.subscriptions.push(remotecompile);

    vscode.commands.registerCommand('StatusBarChange', () => {
        outputChannel.show();
    });
    statusbar.command = "StatusBarChange";
}

// this method is called when your extension is deactivated
export function deactivate() {
}









